### DISCUSSION TITLE
Supported capabilites of JSON Schema

### DISCUSSION THREAD

**Discussion by alemNative** [ory/kratos#1591] [Q&A] [answered]
Hello guys
is there an official documentation what type of fields are supported?

https://json-schema.org/

for example "enum" and "descriptions" fields in props are not outputted. what can i use when i define properties?
can i add own meta objects? how to add a message and so on?

is there something where i see what i can use and what not. with try and error is very time intensive.

also min length, disabled or required fields are not respected and dont apear in the output.

```json
{
  "email": {
....
          "type": "string",
          "format": "email",
          "title": "E-Mail",
          "minLength": 3,
          "ory.sh/kratos": {
            "credentials": {
              "password": {
                "identifier": true
              }
            },
            "verification": {
              "via": "email"
            },
            "recovery": {
              "via": "email"
            }
          }
        },
....
"required": ["email","password"],
```

Output:
```
attributes: Object { <--############# what if select, radio, checkbox?
    disabled: false
​​​    name: "traits.email"
​​​​    type: "email"

​​​​}
messages: Array []
​​​​meta: Object { 
  label: Object { id: 1070002, text: "E-Mail", type: "info" }
  <--######################### can i add something here own?
}
```

I render the form automatic with the response of for example registration initializing. so it would be nice that i can render different field types full automatic even with classes and so on... the meta section should be great

**Comment by vinckr** [comment, marked as answer]
It is best not to store too much data in the Identity Model/Schema. 
It usually should contain only traits that are relevant to that user's identity.
If the data is related more to e.g. your business logic, it would make sense to store it somewhere else ( where it is needed later on).
You also have to keep in mind that identity traits are visible to the user.

If you want to display messages, that is best handled in your front end (depends on what exactly you are looking for).

[This document](https://www.ory.sh/kratos/docs/concepts/identity-data-model#json-schema-vocabulary-extensions) - which you have probably seen - goes into detail on what JSON vocabulary extensions are possible.

At the bottom it says
>There are currently no other extensions supported for Identity Traits. Further fields will be added in future releases!
e.g. https://github.com/ory/kratos/issues/117 or https://github.com/ory/kratos/issues/137.

Required fields not being required is a known issue, please read more about it here: https://github.com/ory/kratos/issues/400 . 

I hope this answers some of your questions, it is a topic touching a lot of points, and please  let me know if I can assist you further with this.

**Reply by alemNative** [reply]
what you mean here:
"You also have to keep in mind that identity traits are visible to the user. "

Is it not possible to mark some traits as private?

**Reply by vinckr** [reply]
I am afraid not, but I am linking the issue that shows what needs to be changed to implement it below.

From our [FAQ](https://www.ory.sh/kratos/docs/next/faq/):

> Currently, traits are visible to the user and also editable by them. It makes sense to have read-only and/or hidden traits for users, but we can not give an exact timing when this will be available. There is an open issue for this feature:
> https://github.com/ory/kratos/issues/47

**Reply by alemNative** [reply]
But doe's that mean that every user can chanage and see his own traits or every user/or even guests  can request any user and see all traits? 

are all products from ory productive ready?

What i need if i wanna manager users locally in my app and just handle the authentications outside? 
but this authenticationsservice has to match also the same identiteies if they comes one time over email login one time over facebook login and so on?

and if i wanna check for example is the logged in user the owner of this and this object (product, blog, whatever) and gave he the pernmission to do this or that?

witch cases you would do with ory stack and wich locally in the app

**Reply by alemNative** [reply]
is ory (and all products) productive ready?

**Reply by zepatrik** [reply]
> But doe's that mean that every user can change and see his own traits or every user/or even guests can request any user and see all traits?

Every user can edit and view their own traits, you will not be able to prevent that at the moment. See above linked issues and documentation for details.

Regarding production-readiness, see https://www.ory.sh/docs/ecosystem/versioning and #918

For follow-up questions, please first search issues/discussions/documentation and then open a new issue/discussion if the question is not resolved by that.
